core/src/saros/net/internal/DataTransferManager.java

```diff
@@ -213,38 +183,44 @@ public synchronized void setServices(int transportMask) {
     setStreamServices();
   }
 
-  /** @deprecated */
   @Override
-  @Deprecated
-  public StreamMode getTransferMode(JID jid) {
-    return getTransferMode(null, jid);
+  public StreamMode getTransferMode(String connectionID, Object address) {
+    IConnection connection = getPacketConnection(connectionID, address);
+    return connection == null ? StreamMode.NONE : connection.getMode();
   }
 
-  @Override
-  public StreamMode getTransferMode(String connectionID, JID jid) {
-    IByteStreamConnection connection = getCurrentConnection(connectionID, jid);
-    return connection == null ? StreamMode.NONE : connection.getMode();
+  public void addPacketConnectionListener(final IPacketConnectionListener listener) {
+    packetConnectionListeners.addIfAbsent(listener);
+  }
+
+  public void removePacketConnectionListener(final IPacketConnectionListener listener) {
+    packetConnectionListeners.remove(listener);
   }
 
-  public void addConnectionListener(final IByteStreamConnectionListener listener) {
-    connectionListeners.addIfAbsent(listener);
+  @Override
+  public void addStreamConnectionListener(final IStreamConnectionListener listener) {
+    streamConnectionListeners.addIfAbsent(listener);
   }
 
-  public void removeConnectionListener(final IByteStreamConnectionListener listener) {
-    connectionListeners.remove(listener);
+  @Override
+  public void removeStreamConnectionListener(final IStreamConnectionListener listener) {
+    streamConnectionListeners.remove(listener);
   }
 
-  private IByteStreamConnection connectInternal(String connectionID, JID peer) throws IOException {
+  private IConnection connectInternal(
+      final String id, final JID address, final boolean isPlainStream) throws IOException {
```

Comment:
This solution using the Boolean to determine the return type is a bit ugly in my opinion. But using generics gets *very* ugly fast as far as I can tell, so this isn't an option either.

Would splitting this method into a stream and non-stream variant be a possibility without creating too much code duplication? The logic is kind of long/complicated, so I don't have a good overview.